src/som/interpreter/processes/SChannel.java

```diff
@@ -70,27 +76,55 @@ public static SChannelInput create(final SynchronousQueue<Object> cell,
 
     private final SynchronousQueue<Object> cell;
     protected final SChannel               channel;
+    private final AtomicInteger            numReads;
 
     public SChannelInput(final SynchronousQueue<Object> cell,
         final SChannel channel) {
       this.cell = cell;
       this.channel = channel;
+      if (VmSettings.ACTOR_TRACING | VmSettings.REPLAY) {
+        numReads = new AtomicInteger(0);
+      } else {
+        numReads = null;
+      }
     }
 
     @TruffleBoundary
-    public Object read() throws InterruptedException {
+    public Object read(final RecordTwoEvent traceRead) throws InterruptedException {
       ObjectTransitionSafepoint.INSTANCE.unregister();
+
+      if (VmSettings.REPLAY) {
+        Activity reader = TracingActivityThread.currentThread().getActivity();
+        ChannelReadRecord crr = (ChannelReadRecord) reader.getNextReplayEvent();
+        assert crr != null : reader;
+        assert channel.getId() == crr.channelId;
+
+        while (numReads.get() != crr.readNo) {
+          Thread.sleep(5); // temporary solution for proof of concept.
+          // maybe use some wait/notify all construct.
+        }
+      }
+
       try {
-        return cell.take();
+        synchronized (this) {
+          if (VmSettings.ACTOR_TRACING) {
+            traceRead.record(channel.getId(), numReads.getAndIncrement());
+          }
+          return cell.take();
+        }
       } finally {
+        if (VmSettings.REPLAY) {
```

Comment:
Would be good to have a comment here why synchronized is not needed.
Or, add a synchronized, and remove the use of the atomic int.
I think, using synchronized might be more consistent, and easier to understand.
Same for the other similar code.

Reply:
added synchronization